aws/helper_functions.py

```diff
@@ -116,19 +71,76 @@ def download_best_model( bucket, s3_model_output, hpo_results, local_directory )
-            full_url = 's3://' + bucket + '/' + path_prefix + obj.key
+            s3_path_to_model = 's3://' + bucket + '/' + path_prefix + filename
             target_bucket.download_file( obj.key, local_filename )
             print(f'Successfully downloaded best model\n'
                   f'> filename: {local_filename}\n'
                   f'> local directory : {local_directory}\n\n'
-                  f'full S3 path : {full_url}')
-        return local_filename
+                  f'full S3 path : {s3_path_to_model}')
+        return local_filename, s3_path_to_model
     
     except Exception as download_error:
         print( f'! Unable to download best model: {download_error}')
-        return None
\ No newline at end of file
+        return None
+    
+def new_job_name_from_config ( dataset_directory, region, code_choice, 
+                               algorithm_choice, cv_folds,
+                               instance_type, trim_limit = 32 ):
+    """ 
+    Build a jobname string that captures the HPO configuration options.
+    This is helpful for intepreting logs and for general book-keeping
+    """
+    job_name = None    
+    try:
+        if dataset_directory in [ '1_year', '3_year', '10_year']:
+            data_choice_str = 'air'
+            validate_region ( region )
+        elif dataset_directory in [ 'NYC_taxi']:
+            data_choice_str = 'nyc'
+            validate_region ( region )
+        else:
+            data_choice_str = 'byo'
+            
+        code_choice_str = code_choice[0] + code_choice[-3:]
+        
+        if 'randomforest' in algorithm_choice.lower() : algorithm_choice_str = 'RF'
+        if 'xgboost' in algorithm_choice.lower() : algorithm_choice_str = 'XGB'    
+        
+        instance_type_str = '-'.join( instance_type.split('.')[1:] )        
+
+        random_8char_str = ''.join( random.choices( uuid.uuid4().hex, k=8 ) )        
+        
+        job_name = f"{data_choice_str}-{code_choice_str}" \
+                    f"-{algorithm_choice_str}-{cv_folds}cv"\
+                    f"-{instance_type_str}-{random_8char_str}"
+        
+        job_name = job_name[:trim_limit]
+        
+        print ( f'generated job name : {job_name}\n')
+        
+    except Exception as error: 
+        print( f'ERROR: unable to generate job name: {error}' )
+    
+    return job_name
+
+def validate_region ( region ):
+    """ Check that the current [compute] region is one of the two regions where the demo data is hosted """
+    if isinstance( region, list ): 
+        region = region[0]
+        
+    if region not in ['us-east-1', 'us-west-2']:
+        raise Exception ( 'Unsupported region based on demo data location,'\
+                          ' please switch to us-east-1 or us-west-2' )
+        
+def get_notebook_path():
+    """ Query the root notebook directory [ once ], 
+        we'll end up placing our best trained model in this location.
+    """
+    try: working_directory
```

Comment:
Sorry, I'm not sure what this code is supposed to do. It looks to me like it will always call and return `os.getcwd()`.

Reply:
This is a remnant from the previous notebook, the idea being that you only check the working directory once (if it already exists skip the check).

Reply:
Ah, I thought that might be what it was trying to do. Unfortunately, this does not actually cache the result. Here's a little snippet to demonstrate:

```python
def incorrect_cache():
    try:
        return result
    except NameError:
        print('Cache miss')
        result = 2 + 2
        return result

def correct_cache():
    try:
        return correct_cache.result
    except AttributeError:
        print('Cache miss')
        correct_cache.result = 2 + 2
        return correct_cache.result


if __name__ == '__main__':
    print('Incorrect cache demonstration:')
    for i in range(10):
        print("{}: {}".format(i, incorrect_cache()))
    print('Correct cache demonstration:')
    for i in range(10):
        print("{}: {}".format(i, correct_cache()))
```

And the output:
```bash
Incorrect cache demonstration:
Cache miss
0: 4
Cache miss
1: 4
Cache miss
2: 4
Cache miss
3: 4
Cache miss
4: 4
Cache miss
5: 4
Cache miss
6: 4
Cache miss
7: 4
Cache miss
8: 4
Cache miss
9: 4
Correct cache demonstration:
Cache miss
0: 4
1: 4
2: 4
3: 4
4: 4
5: 4
6: 4
7: 4
8: 4
9: 4
```
The scope of `working_directory` is local to the function, so referencing it at line 144 is *always* a `NameError`.

Reply:
And I should probably add that the best way to cache is with one of `functools` [caching decorators](https://docs.python.org/3/library/functools.html#functools.cache) unless you need something with more logic to it.